app/org/sagebionetworks/bridge/validators/SurveyAnswerValidator.java

```diff
@@ -176,45 +171,62 @@ private void validateType(Errors errors, DecimalConstraints con, String answer)
-        if (con.getMinLength() != null && answer.length() < con.getMinLength()) {
-            rejectField(errors, "constraints", "%s is shorter than %s characters", answer, con.getMinLength());
-        } else if (con.getMaxLength() != null && answer.length() > con.getMaxLength()) {
-            rejectField(errors, "constraints", "%s is longer than %s characters", answer, con.getMaxLength());
-        }
-        if (StringUtils.isNotBlank(con.getPattern()) && answer != null && !answer.matches(con.getPattern())) {
-            rejectField(errors, "constraints", "%s does not match the regular expression /%s/", answer, con.getPattern());
-        }
-    }
-
     private void validateType(Errors errors, DurationConstraints con, String answer) {
         if (con.getUnit() == null) {
             rejectField(errors, "constraints", "%s does not have a specified unit", answer);
+        } else if (!Unit.DURATION_UNITS.contains(con.getUnit())) {
+            rejectField(errors, "constraints", "%s is not a time unit", answer);
         } else {
             try {
-                answer = new DurationToIntegerConverter().convert(answer, con.getUnit());
-                validateType(errors, (IntegerConstraints)con, answer);
+                String value = new DurationToIntegerConverter().convert(answer, con.getUnit());
+                validateNumericalConstraint(errors, (NumericalConstraints)con, value);
             } catch(IllegalArgumentException e) {
                 rejectField(errors, "constraints", e.getMessage());
             }
         }
     }
     
+    private void validateNumericalConstraint(Errors errors, NumericalConstraints con, String number) {
+        double numberAsDouble = Double.parseDouble(number);
+        if (con.getMinValue() != null && numberAsDouble < con.getMinValue()) {
+            if (con.getUnit() != null) {
+                rejectField(errors, "constraints", "%s is lower than the minimum value of %s %s", number, con.getMinValue(), con.getUnit().name().toLowerCase());
+            } else {
+                rejectField(errors, "constraints", "%s is lower than the minimum value of %s", number, con.getMinValue());
+            }
+        }
+        if (con.getMaxValue() != null && numberAsDouble > con.getMaxValue()) {
+            if (con.getUnit() != null) {
+                rejectField(errors, "constraints", "%s is higher than the maximum value of %s %s", number, con.getMaxValue(), con.getUnit().name().toLowerCase());
+            } else {
+                rejectField(errors, "constraints", "%s is higher than the maximum value of %s", number, con.getMaxValue());
+            }
+        }
+        if (con.getStep() != null) {
+            String result = new BigDecimal(number).remainder(new BigDecimal(con.getStep()), MATH_CONTEXT).toString();
+            if (!result.startsWith("0.0") && !result.equals("0")) { // A simple approach to rounding that works
```

Comment:
Could do straight math on doubles and check that it's less than 0.1.  0.0000001 would be an issue, however.